TestTriggers is flaky

```
--- FAIL: TestTriggers_imageChange-2 (6.15s)
    deploy_trigger_test.go:164: unexpected error: imageStream "test-image-stream" cannot be updated: the object has been modified; please apply your changes to the latest version and try again
FAIL
```

@ironcladlou

```
I1027 23:38:37.491079    2824 create_dockercfg_secrets.go:177] View of ServiceAccount integration/default is not up to date, skipping dockercfg creation
I1027 23:38:37.496554    2824 helper.go:366] UpdateTrackingTags: stream=integration/test-image-stream, updatedTag=latest, updatedImage.dockerImageReference=registry:8080/openshift/test-image@sha256:00000000000000000000000000000001, updatedImage.image=sha256:00000000000000000000000000000001
--- FAIL: TestTriggers_imageChange-2 (6.42s)
	deploy_trigger_test.go:164: unexpected error: imageStream "test-image-stream" cannot be updated: the object has been modified; please apply your changes to the latest version and try again
FAIL
```

--- a/pkg/image/registry/imagestreammapping/rest.go
+++ b/pkg/image/registry/imagestreammapping/rest.go
@@ -5,15 +5,21 @@ import (
 	"k8s.io/kubernetes/pkg/api/errors"
 	"k8s.io/kubernetes/pkg/api/rest"
 	"k8s.io/kubernetes/pkg/api/unversioned"
+	client "k8s.io/kubernetes/pkg/client/unversioned"
 	"k8s.io/kubernetes/pkg/labels"
 	"k8s.io/kubernetes/pkg/runtime"
 	"k8s.io/kubernetes/pkg/util/fielderrors"
+	"k8s.io/kubernetes/pkg/util/wait"
 
 	"github.com/openshift/origin/pkg/image/api"
 	"github.com/openshift/origin/pkg/image/api/validation"
 	"github.com/openshift/origin/pkg/image/registry/image"
 	"github.com/openshift/origin/pkg/image/registry/imagestream"
 )
+
+// maxRetriesOnConflict is the maximum retry count for Create calls which
+// result in resource conflicts.
+const maxRetriesOnConflict = 10
 
 // REST implements the RESTStorage interface in terms of an image registry and
 // image stream registry. It only supports the Create method and is used
@@ -61,7 +67,11 @@ func (s imageStreamMappingStrategy) Validate(ctx kapi.Context, obj runtime.Objec
 	return validation.ValidateImageStreamMapping(mapping)
 }
 
-// Create registers a new image (if it doesn't exist) and updates the specified ImageStream's tags.
+// Create registers a new image (if it doesn't exist) and updates the
+// specified ImageStream's tags. If attempts to update the ImageStream fail
+// with a resource conflict, the update will be retried if the newer
+// ImageStream has no tag diffs from the previous state. If tag diffs are
+// detected, the conflict error is returned.
 func (s *REST) Create(ctx kapi.Context, obj runtime.Object) (runtime.Object, error) {
 	if err := rest.BeforeCreate(Strategy, ctx, obj); err != nil {
 		return nil, err
@@ -90,18 +100,52 @@ func (s *REST) Create(ctx kapi.Context, obj runtime.Object) (runtime.Object, err
 		Image:                image.Name,
 	}
 
-	if !api.AddTagEventToImageStream(stream, tag, next) {
-		// nothing actually changed
-		return &unversioned.Status{Status: unversioned.StatusSuccess}, nil
+	var result *unversioned.Status
+	var updateConflictErr error
+	retryErr := client.RetryOnConflict(wait.Backoff{Steps: maxRetriesOnConflict}, func() error {
+		lastEvent := api.LatestTaggedImage(stream, tag)
+		if !api.AddTagEventToImageStream(stream, tag, next) {
+			// nothing actually changed
+			result = &unversioned.Status{Status: unversioned.StatusSuccess}
+			return nil
+		}
+		api.UpdateTrackingTags(stream, tag, next)
+
+		_, err := s.imageStreamRegistry.UpdateImageStreamStatus(ctx, stream)
+		if err != nil {
+			if !errors.IsConflict(err) {
+				return err
+			}
+			// If the update conflicts, get the latest stream and check for tag
+			// updates. If the latest tag hasn't changed, retry.
+			latestStream, findLatestErr := s.findStreamForMapping(ctx, mapping)
+			if findLatestErr != nil {
+				return findLatestErr
+			}
+			newerEvent := api.LatestTaggedImage(latestStream, tag)
+			if lastEvent == nil || kapi.Semantic.DeepEqual(lastEvent, newerEvent) {
+				// The tag hasn't changed, so try again with the updated stream.
+				stream = latestStream
+				return err
+			}
+			// The tag changed, so return the conflict error back to the client.
+			// This is a little indirect but is necessary because we can't just
+			// return the conflict error from the retry function (since it would
+			// trigger a retry).
+			updateConflictErr = err
+			return nil
+		}
+		result = &unversioned.Status{Status: unversioned.StatusSuccess}
+		return nil
+	})
+
+	if updateConflictErr != nil {
+		return nil, updateConflictErr
 	}
-
-	api.UpdateTrackingTags(stream, tag, next)
-
-	if _, err := s.imageStreamRegistry.UpdateImageStreamStatus(ctx, stream); err != nil {
-		return nil, err
+	if retryErr != nil {
+		return nil, retryErr
 	}
-
-	return &unversioned.Status{Status: unversioned.StatusSuccess}, nil
+	return result, nil
 }
 
 // findStreamForMapping retrieves an ImageStream whose DockerImageRepository matches dockerRepo.

--- a/pkg/image/registry/imagestreammapping/rest.go
+++ b/pkg/image/registry/imagestreammapping/rest.go
@@ -5,7 +5,6 @@ import (
 	"k8s.io/kubernetes/pkg/api/errors"
 	"k8s.io/kubernetes/pkg/api/rest"
 	"k8s.io/kubernetes/pkg/api/unversioned"
-	client "k8s.io/kubernetes/pkg/client/unversioned"
 	"k8s.io/kubernetes/pkg/labels"
 	"k8s.io/kubernetes/pkg/runtime"
 	"k8s.io/kubernetes/pkg/util/fielderrors"
@@ -100,52 +99,39 @@ func (s *REST) Create(ctx kapi.Context, obj runtime.Object) (runtime.Object, err
 		Image:                image.Name,
 	}
 
-	var result *unversioned.Status
-	var updateConflictErr error
-	retryErr := client.RetryOnConflict(wait.Backoff{Steps: maxRetriesOnConflict}, func() error {
+	err = wait.ExponentialBackoff(wait.Backoff{Steps: maxRetriesOnConflict}, func() (bool, error) {
 		lastEvent := api.LatestTaggedImage(stream, tag)
 		if !api.AddTagEventToImageStream(stream, tag, next) {
 			// nothing actually changed
-			result = &unversioned.Status{Status: unversioned.StatusSuccess}
-			return nil
+			return true, nil
 		}
 		api.UpdateTrackingTags(stream, tag, next)
-
 		_, err := s.imageStreamRegistry.UpdateImageStreamStatus(ctx, stream)
-		if err != nil {
-			if !errors.IsConflict(err) {
-				return err
-			}
-			// If the update conflicts, get the latest stream and check for tag
-			// updates. If the latest tag hasn't changed, retry.
-			latestStream, findLatestErr := s.findStreamForMapping(ctx, mapping)
-			if findLatestErr != nil {
-				return findLatestErr
-			}
-			newerEvent := api.LatestTaggedImage(latestStream, tag)
-			if lastEvent == nil || kapi.Semantic.DeepEqual(lastEvent, newerEvent) {
-				// The tag hasn't changed, so try again with the updated stream.
-				stream = latestStream
-				return err
-			}
-			// The tag changed, so return the conflict error back to the client.
-			// This is a little indirect but is necessary because we can't just
-			// return the conflict error from the retry function (since it would
-			// trigger a retry).
-			updateConflictErr = err
-			return nil
+		if err == nil {
+			return true, nil
 		}
-		result = &unversioned.Status{Status: unversioned.StatusSuccess}
-		return nil
+		if !errors.IsConflict(err) {
+			return false, err
+		}
+		// If the update conflicts, get the latest stream and check for tag
+		// updates. If the latest tag hasn't changed, retry.
+		latestStream, findLatestErr := s.findStreamForMapping(ctx, mapping)
+		if findLatestErr != nil {
+			return false, findLatestErr
+		}
+		newerEvent := api.LatestTaggedImage(latestStream, tag)
+		if lastEvent == nil || kapi.Semantic.DeepEqual(lastEvent, newerEvent) {
+			// The tag hasn't changed, so try again with the updated stream.
+			stream = latestStream
+			return false, nil
+		}
+		// The tag changed, so return the conflict error back to the client.
+		return false, err
 	})
-
-	if updateConflictErr != nil {
-		return nil, updateConflictErr
+	if err != nil {
+		return nil, err
 	}
-	if retryErr != nil {
-		return nil, retryErr
-	}
-	return result, nil
+	return &unversioned.Status{Status: unversioned.StatusSuccess}, nil
 }
 
 // findStreamForMapping retrieves an ImageStream whose DockerImageRepository matches dockerRepo.

--- a/pkg/image/registry/imagestreammapping/rest_test.go
+++ b/pkg/image/registry/imagestreammapping/rest_test.go
@@ -1,6 +1,7 @@
 package imagestreammapping
 
 import (
+	"fmt"
 	"net/http"
 	"reflect"
 	"strings"
@@ -9,11 +10,15 @@ import (
 	"github.com/coreos/go-etcd/etcd"
 	kapi "k8s.io/kubernetes/pkg/api"
 	"k8s.io/kubernetes/pkg/api/errors"
+	"k8s.io/kubernetes/pkg/api/unversioned"
+	"k8s.io/kubernetes/pkg/fields"
+	"k8s.io/kubernetes/pkg/labels"
 	"k8s.io/kubernetes/pkg/runtime"
 	kstorage "k8s.io/kubernetes/pkg/storage"
 	etcdstorage "k8s.io/kubernetes/pkg/storage/etcd"
 	"k8s.io/kubernetes/pkg/tools"
 	"k8s.io/kubernetes/pkg/tools/etcdtest"
+	"k8s.io/kubernetes/pkg/watch"
 
 	"github.com/openshift/origin/pkg/api/latest"
 	authorizationapi "github.com/openshift/origin/pkg/authorization/api"
@@ -441,4 +446,193 @@ func TestTrackingTags(t *testing.T) {
 	if e, a := "9999", nonTracking.Image; e != a {
 		t.Errorf("image: expected %s, got %s", e, a)
 	}
+}
+
+// TestCreateRetryUnrecoverable ensures that an attempt to create a mapping
+// using failing registry update calls will return an error.
+func TestCreateRetryUnrecoverable(t *testing.T) {
+	rest := &REST{
+		imageRegistry: &fakeImageRegistry{
+			createImage: func(ctx kapi.Context, image *api.Image) error {
+				return nil
+			},
+		},
+		imageStreamRegistry: &fakeImageStreamRegistry{
+			getImageStream: func(ctx kapi.Context, id string) (*api.ImageStream, error) {
+				return validImageStream(), nil
+			},
+			listImageStreams: func(ctx kapi.Context, selector labels.Selector) (*api.ImageStreamList, error) {
+				s := validImageStream()
+				return &api.ImageStreamList{Items: []api.ImageStream{*s}}, nil
+			},
+			updateImageStreamStatus: func(ctx kapi.Context, repo *api.ImageStream) (*api.ImageStream, error) {
+				return nil, errors.NewServiceUnavailable("unrecoverable error")
+			},
+		},
+	}
+	obj, err := rest.Create(kapi.NewDefaultContext(), validNewMappingWithName())
+	if err == nil {
+		t.Errorf("expected an error")
+	}
+	if obj != nil {
+		t.Fatalf("expected a nil result")
+	}
+}
+
+// TestCreateRetryConflictNoTagDiff ensures that attempts to create a mapping
+// that result in resource conflicts that do NOT include tag diffs causes the
+// create to be retried successfully.
+func TestCreateRetryConflictNoTagDiff(t *testing.T) {
+	firstUpdate := true
+	rest := &REST{
+		imageRegistry: &fakeImageRegistry{
+			createImage: func(ctx kapi.Context, image *api.Image) error {
+				return nil
+			},
+		},
+		imageStreamRegistry: &fakeImageStreamRegistry{
+			getImageStream: func(ctx kapi.Context, id string) (*api.ImageStream, error) {
+				stream := validImageStream()
+				stream.Status = api.ImageStreamStatus{
+					Tags: map[string]api.TagEventList{
+						"latest": {Items: []api.TagEvent{{DockerImageReference: "localhost:5000/someproject/somerepo:original"}}},
+					},
+				}
+				return stream, nil
+			},
+			updateImageStreamStatus: func(ctx kapi.Context, repo *api.ImageStream) (*api.ImageStream, error) {
+				// For the first update call, return a conflict to cause a retry of an
+				// image stream whose tags haven't changed.
+				if firstUpdate {
+					firstUpdate = false
+					return nil, errors.NewConflict("ImageStream", repo.Name, fmt.Errorf("resource modified"))
+				}
+				return repo, nil
+			},
+		},
+	}
+	obj, err := rest.Create(kapi.NewDefaultContext(), validNewMappingWithName())
+	if err != nil {
+		t.Errorf("unexpected error: %v", err)
+	}
+	if obj == nil {
+		t.Fatalf("expected a result")
+	}
+}
+
+// TestCreateRetryConflictTagDiff ensures that attempts to create a mapping
+// that result in resource conflicts that DO contain tag diffs causes the
+// conflict error to be returned.
+func TestCreateRetryConflictTagDiff(t *testing.T) {
+	firstGet := true
+	firstUpdate := true
+	rest := &REST{
+		imageRegistry: &fakeImageRegistry{
+			createImage: func(ctx kapi.Context, image *api.Image) error {
+				return nil
+			},
+		},
+		imageStreamRegistry: &fakeImageStreamRegistry{
+			getImageStream: func(ctx kapi.Context, id string) (*api.ImageStream, error) {
+				// For the first get, return a stream with a latest tag pointing to "original"
+				if firstGet {
+					firstGet = false
+					stream := validImageStream()
+					stream.Status = api.ImageStreamStatus{
+						Tags: map[string]api.TagEventList{
+							"latest": {Items: []api.TagEvent{{DockerImageReference: "localhost:5000/someproject/somerepo:original"}}},
+						},
+					}
+					return stream, nil
+				}
+				// For subsequent gets, return a stream with the latest tag changed to "newer"
+				stream := validImageStream()
+				stream.Status = api.ImageStreamStatus{
+					Tags: map[string]api.TagEventList{
+						"latest": {Items: []api.TagEvent{{DockerImageReference: "localhost:5000/someproject/somerepo:newer"}}},
+					},
+				}
+				return stream, nil
+			},
+			updateImageStreamStatus: func(ctx kapi.Context, repo *api.ImageStream) (*api.ImageStream, error) {
+				// For the first update, return a conflict so that the stream
+				// get/compare is retried.
+				if firstUpdate {
+					firstUpdate = false
+					return nil, errors.NewConflict("ImageStream", repo.Name, fmt.Errorf("resource modified"))
+				}
+				return repo, nil
+			},
+		},
+	}
+	obj, err := rest.Create(kapi.NewDefaultContext(), validNewMappingWithName())
+	if err == nil {
+		t.Fatalf("expected an error")
+	}
+	if !errors.IsConflict(err) {
+		t.Errorf("expected a conflict error, got %v", err)
+	}
+	if obj != nil {
+		t.Fatalf("expected a nil result")
+	}
+}
+
+type fakeImageRegistry struct {
+	listImages  func(ctx kapi.Context, selector labels.Selector) (*api.ImageList, error)
+	getImage    func(ctx kapi.Context, id string) (*api.Image, error)
+	createImage func(ctx kapi.Context, image *api.Image) error
+	deleteImage func(ctx kapi.Context, id string) error
+	watchImages func(ctx kapi.Context, label labels.Selector, field fields.Selector, resourceVersion string) (watch.Interface, error)
+}
+
+func (f *fakeImageRegistry) ListImages(ctx kapi.Context, selector labels.Selector) (*api.ImageList, error) {
+	return f.listImages(ctx, selector)
+}
+func (f *fakeImageRegistry) GetImage(ctx kapi.Context, id string) (*api.Image, error) {
+	return f.getImage(ctx, id)
+}
+func (f *fakeImageRegistry) CreateImage(ctx kapi.Context, image *api.Image) error {
+	return f.createImage(ctx, image)
+}
+func (f *fakeImageRegistry) DeleteImage(ctx kapi.Context, id string) error {
+	return f.deleteImage(ctx, id)
+}
+func (f *fakeImageRegistry) WatchImages(ctx kapi.Context, label labels.Selector, field fields.Selector, resourceVersion string) (watch.Interface, error) {
+	return f.watchImages(ctx, label, field, resourceVersion)
+}
+
+type fakeImageStreamRegistry struct {
+	listImageStreams        func(ctx kapi.Context, selector labels.Selector) (*api.ImageStreamList, error)
+	getImageStream          func(ctx kapi.Context, id string) (*api.ImageStream, error)
+	createImageStream       func(ctx kapi.Context, repo *api.ImageStream) (*api.ImageStream, error)
+	updateImageStream       func(ctx kapi.Context, repo *api.ImageStream) (*api.ImageStream, error)
+	updateImageStreamSpec   func(ctx kapi.Context, repo *api.ImageStream) (*api.ImageStream, error)
+	updateImageStreamStatus func(ctx kapi.Context, repo *api.ImageStream) (*api.ImageStream, error)
+	deleteImageStream       func(ctx kapi.Context, id string) (*unversioned.Status, error)
+	watchImageStreams       func(ctx kapi.Context, label labels.Selector, field fields.Selector, resourceVersion string) (watch.Interface, error)
+}
+
+func (f *fakeImageStreamRegistry) ListImageStreams(ctx kapi.Context, selector labels.Selector) (*api.ImageStreamList, error) {
+	return f.listImageStreams(ctx, selector)
+}
+func (f *fakeImageStreamRegistry) GetImageStream(ctx kapi.Context, id string) (*api.ImageStream, error) {
+	return f.getImageStream(ctx, id)
+}
+func (f *fakeImageStreamRegistry) CreateImageStream(ctx kapi.Context, repo *api.ImageStream) (*api.ImageStream, error) {
+	return f.createImageStream(ctx, repo)
+}
+func (f *fakeImageStreamRegistry) UpdateImageStream(ctx kapi.Context, repo *api.ImageStream) (*api.ImageStream, error) {
+	return f.updateImageStream(ctx, repo)
+}
+func (f *fakeImageStreamRegistry) UpdateImageStreamSpec(ctx kapi.Context, repo *api.ImageStream) (*api.ImageStream, error) {
+	return f.updateImageStreamSpec(ctx, repo)
+}
+func (f *fakeImageStreamRegistry) UpdateImageStreamStatus(ctx kapi.Context, repo *api.ImageStream) (*api.ImageStream, error) {
+	return f.updateImageStreamStatus(ctx, repo)
+}
+func (f *fakeImageStreamRegistry) DeleteImageStream(ctx kapi.Context, id string) (*unversioned.Status, error) {
+	return f.deleteImageStream(ctx, id)
+}
+func (f *fakeImageStreamRegistry) WatchImageStreams(ctx kapi.Context, label labels.Selector, field fields.Selector, resourceVersion string) (watch.Interface, error) {
+	return f.watchImageStreams(ctx, label, field, resourceVersion)
 }

